### PR TITLE
[GTK] Slow scroll adjustment when using a mouse wheel

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### 645e981e523ddb5f43d5b53a545d320ef0f38c25
<pre>
[GTK] Slow scroll adjustment when using a mouse wheel
<a href="https://bugs.webkit.org/show_bug.cgi?id=168300">https://bugs.webkit.org/show_bug.cgi?id=168300</a>

Reviewed by Michael Catanzaro.

Use a different scroll step depending on the scrollable area size,
matching what GTK4 does.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::updateScrollbarSteps):
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::updateScrollbarSteps):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::pixelsPerLineStep):
* Source/WebCore/platform/Scrollbar.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollbarSteps):
* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/gtk/NativeWebWheelEventGtk.cpp:
* Source/WebKit/Shared/gtk/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebWheelEvent): Deleted.
* Source/WebKit/Shared/gtk/WebEventFactory.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseScrollEvent):
(handleScroll):
(webkitWebViewBaseHandleWheelEvent): Deleted.

Canonical link: <a href="https://commits.webkit.org/261961@main">https://commits.webkit.org/261961@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c0963705466d83394d28281d0581d0aa8e7bb42a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/91 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/102 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/130 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/84 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/95 "10 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/87 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/102 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100 "11 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/96 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/84 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/82 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/100 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by carlosgcampos** [comment]
The patch has a `WEBKIT_SCROLL_STEP` env var just to be able to try different values without having to compile. But if we all are happy with the GTK4 approach we can just remove the env var.

File: Source/WebCore/platform/Scrollbar.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/Scrollbar.cpp]: This is great for testing, but let's take this out before landing to ensure applications do not try to use it.

File: Source/WebCore/platform/Scrollbar.h
Comment (mcatanzaro) [inline Source/WebCore/platform/Scrollbar.h:50]: Can you make this one private to ensure callers have to pass the int parameter, or is it still used elsewhere? (Removing the option to not pass the parameter would discourage misuse.) Actually, it's probably cleaner if we have just one function here. Also, I think we should still change this to default to 120 (at least for Windows and WPE) and use 40 only for Cocoa. So I would do something like:

```
int Scrollbar::pixelsPerLineStep(int viewWidthOrHeight)
{
#if PLATFORM(GTK)
    if (const char* scrollStep = getenv("WEBKIT_SCROLL_STEP")) {
        int value;
        if (sscanf(scrollStep, "%d", &value) == 1 && value > 0)
            return value;
    }

    if (viewWidthOrHeight > 0)
        return std::pow(viewWidthOrHeight, 2. / 3.);
#elif PLATFORM(COCOA)
    UNUSED_PARAM(viewWidthOrHeight);
    return 40;
#else
    UNUSED_PARAM(viewWidthOrHeight);
    return 120;
#endif
}
```

**Review by mcatanzaro — APPROVED** [review]
So this is a big improvement and I think it will resolve user complaints. The scrolling is faster than our current scroll speed at all reasonable window sizes, although not as fast as Chrome unless you have a large window. Nice.

I added some debug to print the computed scroll step, and was surprised to see that the viewWidthOrHeight keeps changing during every scroll step even when the web view size stays the same:

```
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
pixelsPerLineStep: widthOrHeight=1064 step=104
pixelsPerLineStep: widthOrHeight=811 step=86
```

This prints on pages with only vertical scrollbars and no horizontal scrollbars, so something is surely wrong there, although it's probably a bug elsewhere.

**Comment by mcatanzaro** [comment]
> Note this is unlikely to survive EWS as it will affect layout test results.

Looks like only one test needs to be rebaselined:

```
--- /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/imported/w3c/web-platform-tests/css/css-scroll-snap/input/snap-area-overflow-boundary-expected.txt
+++ /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/imported/w3c/web-platform-tests/css/css-scroll-snap/input/snap-area-overflow-boundary-actual.txt
@@ -1,6 +1,6 @@
 Header 1Footer 1
 Header 2Footer 2
 
-FAIL Keyboard scrolling with vertical snap-area overflow assert_equals: End boundary of snap-area is valid snap target expected 39 but got 449
+FAIL Keyboard scrolling with vertical snap-area overflow assert_equals: End boundary of snap-area is valid snap target expected 53 but got 463
 FAIL Mouse-wheel scrolling with vertical snap-area overflow promise_test: Unhandled rejection with value: object "Error: Unknown source type "wheel"."
 ```

Although this is a WPT test so it's not really supposed to have platform-specific results, is it? Oh well. Change looks fine.

File: Source/WebCore/platform/Scrollbar.cpp
Comment (carlosgcampos) [reply Source/WebCore/platform/Scrollbar.cpp]: Yes, that's the plan.

**Comment by carlosgcampos** [comment]
> This prints on pages with only vertical scrollbars and no horizontal scrollbars, so something is surely wrong there, although it's probably a bug elsewhere.

I don't think there's any bug, we just compute both x and y step on every scroll event.

File: Source/WebCore/platform/Scrollbar.h
Comment (carlosgcampos) [reply Source/WebCore/platform/Scrollbar.h:50]: This is used in other places where we don't want to pass the view size. I don't think we should change the behavior of other ports in this patch, for wpe we should probably do the same than gtk, but in a follow up after discussing with other wpe maintainers.

**Comment by carlosgcampos** [comment]
> Note this is unlikely to survive EWS as it will affect layout test results.

It shouldn't because synthesized events are still using 40.

**Pull request merged**
